multitenancy/decision_manager.go

```diff
@@ -0,0 +1,199 @@
+// Note: place holder for future, this is to protect Value Transfer between accounts.
+type AccountAccessDecisionManager interface {
+	IsAuthorized(ctx context.Context, authToken *proto.PreAuthenticatedAuthenticationToken, attr *AccountStateSecurityAttribute) (bool, error)
+}
+
+// ContractAccessDecisionManager performs authorization checks for contract
+// based on what is entitled in the proto.PreAuthenticatedAuthenticationToken
+// and what is asked in ContractSecurityAttribute list.
+type ContractAccessDecisionManager interface {
+	IsAuthorized(ctx context.Context, authToken *proto.PreAuthenticatedAuthenticationToken, attributes []*ContractSecurityAttribute) (bool, error)
+}
+
+type DefaultContractAccessDecisionManager struct {
+}
+
+// IsAuthorized performs authorization check for each security attribute against
+// the granted access inside the pre-authenticated access token.
+//
+// All security attributes must pass.
+func (cm *DefaultContractAccessDecisionManager) IsAuthorized(ctx context.Context, authToken *proto.PreAuthenticatedAuthenticationToken, attributes []*ContractSecurityAttribute) (bool, error) {
+	matchCount := 0
+	if len(attributes) == 0 {
+		return false, nil
+	}
+	for _, attr := range attributes {
+		query := url.Values{}
+		switch attr.Visibility {
+		case VisibilityPublic:
+			switch attr.Action {
+			case ActionRead, ActionWrite, ActionCreate:
+				if (attr.To == common.Address{}) {
+					query.Set(QueryOwnedEOA, toHexAddress(attr.From))
+				} else {
+					query.Set(QueryOwnedEOA, toHexAddress(attr.To))
+				}
+			}
+		case VisibilityPrivate:
+			switch attr.Action {
+			case ActionRead, ActionWrite:
+				if (attr.To == common.Address{}) {
+					query.Set(QueryOwnedEOA, toHexAddress(attr.From))
+				} else {
+					query.Set(QueryOwnedEOA, toHexAddress(attr.To))
+				}
+				for _, tm := range attr.Parties {
+					query.Add(QueryFromTM, tm)
+				}
+			case ActionCreate:
+				query.Set(QueryFromTM, attr.PrivateFrom)
+			}
+		}
+		// construct request permission identifier
+		request, err := url.Parse(fmt.Sprintf("%s://%s/%s/%s?%s", attr.Visibility, strings.ToLower(attr.From.Hex()), attr.Action, "contracts", query.Encode()))
+		if err != nil {
+			return false, err
+		}
+		// compare the contract security attribute with the consolidate list
+		for _, granted := range authToken.GetAuthorities() {
+			pi, err := url.Parse(granted.GetRaw())
+			if err != nil {
+				continue
+			}
+			granted := pi.String()
+			ask := request.String()
+			isMatched := match(attr, request, pi)
+			log.Debug("Checking contract access", "passed", isMatched, "granted", granted, "ask", ask)
+			if isMatched {
+				matchCount++
+				break
+			}
+		}
+	}
+	return matchCount == len(attributes), nil
+}
+
+func toHexAddress(a common.Address) string {
+	if (a == common.Address{}) {
+		return AnyEOAAddress
+	}
+	return strings.ToLower(a.Hex())
+}
+
+func match(attr *ContractSecurityAttribute, ask, granted *url.URL) bool {
```

Comment:
Commenting here about the general format of the scopes
- I think it would be great to be able to pass multiple account address as part of a single scope (if I understand correctly currently it is only possible pass either one address or a wildcard giving access to all accounts thus forcing to create multiple similar scopes when requiring permissions on multiple addresses which is not convenient)
- I am a bit disturbed by the read/write access control which sounds redundant with permissioning while multi-tenancy is mainly meant to provide state access control. In the end, I feel like we are bringing in another variable for users to configure increasing complexity for users. In the end if we anticipate that users will always use `_` for read/write access then I am not certain we need to make this configurable at the scope level (this is a personal opinion that I am glad to discuss, I am usually more in favor of doing simple so we avoid support questions in the future)

Reply:
The read/write access for multi-tenancy is meant for a specific node(and it's state) that is linked to an authorization server, not for the entire network. The permissioning model is for read/write to the network, not any specific node and it's also visible to the entire network.